### PR TITLE
Add SBOM manifest to generated artifacts

### PR DESCRIPTION
**Pull request by michaelpeng36**
### Issue describing the changes in this PR

Adds SBOM generation for release builds. Other changes include:
- Creating an IsReleaseBuild environmental variable to easily identify scenarios where generating SBOM manifests and .msi files are necessary
- Adding a `pipelineUtilities.psm1` file containing SDK dependencies and functions needed by the build pipeline
- Dividing .nupkg generation into a publish and a pack step so SBOM generation may occur between them

### Pull request checklist

* [x] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
  * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
  * [x] Otherwise: Backport tracked by issue/PR #2871 
* [ ] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -23,6 +23,17 @@ jobs:
     APPVEYOR_REPO_BRANCH: $[coalesce(variables['System.PullRequest.TargetBranch'], variables['Build.SourceBranchName'])]
     APPVEYOR_REPO_COMMIT: $(Build.SourceVersion)
   steps:
+  - pwsh: |
+      $isReleaseBuild = $false
+      if ($env:BuildSourceBranchName -like "release_4.0-*")
```
Comment (Francisco-Gamino) [inline azure-pipelines.yml:28]: Should this `"release_4.0-*"` be a pipeline variable? This way we can change the target branch for a release without having to make a code change.

File: build.ps1 [hunk 2]
```diff
@@ -19,6 +19,10 @@ if ($env:IntegrationBuildNumber)
 
     $buildCommand = "dotnet run --integrationTests"
 }
+elseif ($env:IsReleaseBuild)
```
Comment (Francisco-Gamino) [inline build.ps1:22]: Do we only want to include a SBOM for release builds?

File: build/BuildSteps.cs [hunk 3]
```diff
@@ -505,6 +507,67 @@ private static string CurrentVersion
...
+                // Generate the SBOM manifest
+                Shell.Run("dotnet",
+                    $"{Settings.SBOMManifestToolPath} generate -PackageName {packageName} -BuildDropPath {buildPath}"
+                    + $" -BuildComponentPath {buildPath} -Verbosity Information -t {telemetryFilePath}");
+            }
+        }
+
+        public static void DotnetPublishForNupkg()
+        {
+            // By default, this publishes to the /bin/Release/net6.0/publish
```
Comment (Francisco-Gamino) [inline build/BuildSteps.cs:536]: Please update the comment path to `/bin/Release/$framework/publish`

File: build/BuildSteps.cs [hunk 4]
```diff
@@ -505,6 +507,67 @@ private static string CurrentVersion
...
+                                $"/p:NoWorkers=\"true\" " +
+                                $"/p:CommitHash=\"{Settings.CommitId}\" " +
+                                (string.IsNullOrEmpty(Settings.IntegrationBuildNumber) ? string.Empty : $"/p:IntegrationBuildNumber=\"{Settings.IntegrationBuildNumber}\" ") +
+                                $"-c Release");
+        }
+
+        public static void GenerateSBOMManifestForNupkg()
+        {
+            Directory.CreateDirectory(Settings.SBOMManifestTelemetryDir);
+            // Generate the SBOM manifest for the .nupkg file
```
Comment (Francisco-Gamino) [inline build/BuildSteps.cs:548]: I don't think we need either of these two comments.

File: build/Program.cs [hunk 5]
```diff
@@ -29,8 +29,12 @@ static void Main(string[] args)
                 .Then(TestPreSignedArtifacts, skip: !args.Contains("--ci"))
                 .Then(CopyBinariesToSign, skip: !args.Contains("--ci"))
                 .Then(Test)
+                .Then(GenerateSBOMManifestForZips, skip: !args.Contains("addSBOM"))
```
Comment (Francisco-Gamino) [inline build/Program.cs:32]: This should be `!args.Contains("--addSBOM")` for consistency.

File: build/Settings.cs [hunk 6]
```diff
@@ -25,6 +25,7 @@ private static string config(string @default = null, [CallerMemberName] string k
         public const string DotnetProjectTemplatesVersion = "4.0.1964";
         public const string TemplateJsonVersion = "3.1.1648";
 
+        public static readonly string SBOMManifestToolPath = Path.GetFullPath("../ManifestTool/Microsoft.ManifestTool.dll");
```
Comment (Francisco-Gamino) [inline build/Settings.cs:28]: nit: Please add an empty line between `SBOMManifestToolPath` and `SrcProjectPath`.

File: generateMsiFiles.ps1 [hunk 7]
```diff
@@ -45,6 +45,7 @@ if ($MsiGenBranches -Contains $env:APPVEYOR_REPO_BRANCH -or $CommitMessage -Cont
         $fragmentPath = "$buildDir\$fragmentName.wxs"
         $msiPath = "$artifactsPath\$msiName.msi"
 
+
```
Comment (Francisco-Gamino) [inline generateMsiFiles.ps1:48]: nit: Please remove the extra line.

File: helper.psm1 [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    {
+        throw "$MANIFESTOOL_DIRECTORY does not contain '$DLL_NAME'"
+    }
+
+    Write-Host 'Done.'
+
+    return $MANIFEST_TOOL_PATH
+}
+
+$IsWindowsEnv = [RuntimeInformation]::IsOSPlatform([OSPlatform]::Windows)
```
Comment (Francisco-Gamino) [inline helper.psm1:50]: This type is not available in `pwsh`. Instead you can use `$IsWindows`, which is available in PowerShell 6 or higher. In addition, please add this `#Requires -Version 6.0` to the top of the script. For a full example, please see https://github.com/Azure/azure-functions-powershell-worker/blob/dev/build.ps1#L40.

File: helper.psm1 [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    }
+}
+
+function AddLocalDotnetDirPath {
+    $LocalDotnetDirPath = if ($IsWindowsEnv) { "$env:LocalAppData\Microsoft\dotnet" } else { "$env:HOME/.dotnet" }
+    if (($env:PATH -split [IO.Path]::PathSeparator) -notcontains $LocalDotnetDirPath) {
+        $env:PATH = $LocalDotnetDirPath + [IO.Path]::PathSeparator + $env:PATH
+    }
+}
+function Find-Dotnet
```
Comment (Francisco-Gamino) [inline helper.psm1:72]: nit: Please add a space between 71 and 72.

File: helper.psm1 [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        $firstAcceptable = $installedDotnetSdks |
+                                Where-Object { $_.StartsWith("$majorMinorVersion.") } |
+                                Where-Object { [System.Management.Automation.SemanticVersion]::new($_) -ge [System.Management.Automation.SemanticVersion]::new($minimalVersion) } |
+                                Select-Object -First 1
+        if (-not $firstAcceptable) {
+            throw "Cannot find the dotnet SDK for .NET Core $majorMinorVersion. Version $minimalVersion or higher is required. Please specify '-Bootstrap' to install build dependencies."
+        }
+    }
+}
+function Install-Dotnet {
```
Comment (Francisco-Gamino) [inline helper.psm1:89]: Please add a space.

**Review by Francisco-Gamino — CHANGES_REQUESTED** [review]
Thank you for sending this out! It looks great overall. I left a few comments.

File: build/Program.cs [hunk 11]
```diff
@@ -15,24 +15,28 @@ static void Main(string[] args)
...
                 .Then(AddDistLib)
                 .Then(AddTemplatesNupkgs)
                 .Then(AddTemplatesJson)
-                .Then(AddGoZip)
-                .Then(TestPreSignedArtifacts, skip: !args.Contains("--ci"))
-                .Then(CopyBinariesToSign, skip: !args.Contains("--ci"))
-                .Then(Test)
+                //.Then(AddGoZip)
+                //.Then(TestPreSignedArtifacts, skip: !args.Contains("--ci"))
+                //.Then(CopyBinariesToSign, skip: !args.Contains("--ci"))
```
Comment (soninaren) [inline build/Program.cs:30]: Please make sure this does not get checked in :).

File: build.ps1 (hunk 2 above)
Comment (michaelpeng36) [reply build.ps1:22]: Yes, we will only be including an SBOM for release builds.

File: azure-pipelines.yml [hunk 12]
```diff
@@ -38,15 +49,16 @@ jobs:
...
-        # Official release versions can be found at: https://dotnet.microsoft.com/download/dotnet/6.0
-        # Newer versions can be found at: https://github.com/dotnet/installer#installers-and-binaries
-        ./dotnet-install.ps1 -InstallDir 'C:\Program Files\dotnet' -Verbose -Version 6.0.100
+  - pwsh: |
+      Import-Module "$PSScriptRoot/helper.psm1" -Force
+      Install-Dotnet
+    displayName: 'Install .NET 6.0 and 3.1'
+  - pwsh: |
+      Import-Module "$PSScriptRoot/helper.psm1" -Force
+      Install-SBOMUtil $env:SBOMUtilSASUrl
```
Comment (Francisco-Gamino) [inline azure-pipelines.yml:58]: Please include the parameter name. E.g., `Install-SBOMUtil -SBOMUtilSASUrl $env:SBOMUtilSASUrl`

File: azure-pipelines.yml [hunk 13]
```diff
@@ -38,15 +49,16 @@ jobs:
...
-    displayName: 'Install .NET 6'
-    inputs:
-      targetType: 'inline'
-      script: |
-        Invoke-WebRequest 'https://dot.net/v1/dotnet-install.ps1' -OutFile 'dotnet-install.ps1'
-        # Official release versions can be found at: https://dotnet.microsoft.com/download/dotnet/6.0
-        # Newer versions can be found at: https://github.com/dotnet/installer#installers-and-binaries
-        ./dotnet-install.ps1 -InstallDir 'C:\Program Files\dotnet' -Verbose -Version 6.0.100
+  - pwsh: |
+      Import-Module "$PSScriptRoot/helper.psm1" -Force
```
Comment (Francisco-Gamino) [inline azure-pipelines.yml:53]: We should change this module name `helper.psm1` to something more descriptive.

**Review by Francisco-Gamino — APPROVED** [review]
A couple of minor comments. Otherwise, LGTM.

**Pull request merged**